arbnode/inbox_reader.go

```diff
@@ -159,6 +159,46 @@ func (r *InboxReader) Start(ctxIn context.Context) error {
 	return nil
 }
 
+// assumes l1block is recent so we could do a simple-search from the end
+func (r *InboxReader) recentL1BlockToMsg(ctx context.Context, l1block uint64) (arbutil.MessageIndex, error) {
+	batch, err := r.tracker.GetBatchCount()
+	if err != nil {
+		return 0, err
+	}
+	for {
+		if ctx.Err() != nil {
+			return 0, ctx.Err()
+		}
+		if batch == 0 {
+			return 0, nil
+		}
+		batch -= 1
+		meta, err := r.tracker.GetBatchMetadata(batch)
+		if err != nil {
+			return 0, err
+		}
+		if meta.L1Block < l1block {
+			return meta.MessageCount, nil
+		}
+	}
+}
+
+func (r *InboxReader) GetSafeMsg(ctx context.Context) (arbutil.MessageIndex, error) {
+	l1block, err := r.l1Reader.LatestSafeBlockNr(ctx)
+	if err != nil {
+		return 0, err
+	}
+	return r.recentL1BlockToMsg(ctx, l1block)
+}
+
+func (r *InboxReader) GetFinalizedMsg(ctx context.Context) (arbutil.MessageIndex, error) {
```

Comment:
We should consider calling these methods `GetFinalizedMsgCount` (same for the safe one) to clarify that the message at the returned index is not finalized. This is a bit different than the finalized block, where that block number is finalized.